programs/mango-v4/src/instructions/openbook_v2_place_order.rs

```diff
@@ -0,0 +1,538 @@
+    // when they cross the oracle
+    //
+    let openbook = account.openbook_v2_orders_mut(openbook_market.market_index)?;
+    if !before_had_bids {
+        // The 0 state means uninitialized/no value
+        openbook.highest_placed_bid_inv = 0.0;
+        openbook.lowest_placed_bid_inv = 0.0
+    }
+    if !before_had_asks {
+        openbook.lowest_placed_ask = 0.0;
+        openbook.highest_placed_ask = 0.0;
+    }
+    // in the normal quote per base units
+    let limit_price =
+        order.max_quote_lots_including_fees as f64 * quote_lot_size as f64 / base_lot_size as f64;
+
+    let new_order_on_book = after_oo_free_slots != before_oo_free_slots;
+    if new_order_on_book {
+        match order.side {
+            OpenbookV2Side::Ask => {
+                openbook.lowest_placed_ask = if openbook.lowest_placed_ask == 0.0 {
+                    limit_price
+                } else {
+                    openbook.lowest_placed_ask.min(limit_price)
+                };
+                openbook.highest_placed_ask = if openbook.highest_placed_ask == 0.0 {
+                    limit_price
+                } else {
+                    openbook.highest_placed_ask.max(limit_price)
+                }
+            }
+            OpenbookV2Side::Bid => {
+                // in base per quote units, to avoid a division in health
+                let limit_price_inv = 1.0 / limit_price;
+                openbook.highest_placed_bid_inv = if openbook.highest_placed_bid_inv == 0.0 {
+                    limit_price_inv
+                } else {
+                    // the highest bid has the lowest _inv value
+                    openbook.highest_placed_bid_inv.min(limit_price_inv)
+                };
+                openbook.lowest_placed_bid_inv = if openbook.lowest_placed_bid_inv == 0.0 {
+                    limit_price_inv
+                } else {
+                    // lowest bid has max _inv value
+                    openbook.lowest_placed_bid_inv.max(limit_price_inv)
+                }
+            }
+        }
+    }
+
+    emit!(OpenbookV2OpenOrdersBalanceLog {
+        mango_group: ctx.accounts.group.key(),
+        mango_account: ctx.accounts.account.key(),
+        market_index: openbook_market.market_index,
+        base_token_index: openbook_market.base_token_index,
+        quote_token_index: openbook_market.quote_token_index,
+        base_total: after_oo.native_base_total(),
+        base_free: after_oo.native_base_free(),
+        quote_total: after_oo.native_quote_total(),
+        quote_free: after_oo.native_quote_free(),
+        referrer_rebates_accrued: after_oo.native_rebates(),
+    });
+
+    ctx.accounts.payer_vault.reload()?;
+    let after_vault = ctx.accounts.payer_vault.amount;
+
+    // Placing an order cannot increase vault balance
+    require_gte!(before_vault, after_vault);
+
+    let mut payer_bank = ctx.accounts.payer_bank.load_mut()?;
+    let mut receiver_bank = ctx.accounts.receiver_bank.load_mut()?;
```

Comment:
need to verify that these accounts are the right ones, check the token_index for both (anchor this as `#4` and mention it in the comment on payer_bank and receiver_bank in the accounts struct)

Technically these indexes are assert'ed in update_bank_potential_tokens() but I'd prefer if the check was explicit here.

Reply:
does the check in `#3` not cover this by checking against the market token indexes?

Reply:
no, the checks at `#3` don't really check what they should:
```
    let (receiver_bank, receiver_bank_oracle) =
        retriever.bank_and_oracle(&group_key, receiver_active_index, receiver_token_index)?;
    require_eq!(receiver_bank.token_index, receiver_token_index);
```
This just checks that `bank_and_oracle()` did its job (which read the bank from the health account list) (and this require_eq should likely be removed). Replace that check with `ctx.accounts.receiver_bank.load()?.token_index == receiver_token_index`, i.e. that the explicitly-passed bank is for the right token.